common/auth/tls_config_helper.go

```diff
@@ -95,3 +99,87 @@ func tlsCN(state tls.ConnectionState) string {
 	}
 	return state.PeerCertificates[0].Subject.CommonName
 }
+
+func NewTLSConfig(temporalTls *TLS) (*tls.Config, error) {
+	if temporalTls == nil || !temporalTls.Enabled {
+		return nil, nil
+	}
+	if temporalTls.CertData != "" && temporalTls.CertFile != "" {
+		return nil, errors.New("only one of certData or certFile properties should be specified")
+	}
+
+	if temporalTls.KeyData != "" && temporalTls.KeyFile != "" {
+		return nil, errors.New("only one of keyData or keyFile properties should be specified")
+	}
+
+	if temporalTls.CaData != "" && temporalTls.CaFile != "" {
+		return nil, errors.New("only one of caData or caFile properties should be specified")
+	}
+
+	tlsConfig := &tls.Config{
+		InsecureSkipVerify: !temporalTls.EnableHostVerification,
+	}
+	if temporalTls.ServerName != "" {
+		tlsConfig.ServerName = temporalTls.ServerName
+	}
+
+	// Load CA cert
+	var caBytes []byte
+	var err error
+	if temporalTls.CaFile != "" {
+		caBytes, err = os.ReadFile(temporalTls.CaFile)
+		if err != nil {
+			return nil, fmt.Errorf("error reading client ca file %w", err)
+		}
+	} else if temporalTls.CaData != "" {
+		caBytes, err = base64.StdEncoding.DecodeString(temporalTls.CaData)
+		if err != nil {
+			return nil, fmt.Errorf("error decoding client ca data %w", err)
+		}
+	}
+	if len(caBytes) > 0 {
+		caCertPool := x509.NewCertPool()
+		if !caCertPool.AppendCertsFromPEM(caBytes) {
```

Comment:
We should validate the certificate here using [`x509.ParseCertificate`](https://pkg.go.dev/crypto/x509@go1.21.4#ParseCertificate) so that we can return an error in case its malformed. This is what AppendCertsFromPEM does internally but it swallows the error